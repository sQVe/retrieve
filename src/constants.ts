import { Container } from './types'

export const defaultContainer: Container = {
  url: '',
  init: {
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    method: 'GET',
  },
  options: { resolveAs: 'JSON' },
}
