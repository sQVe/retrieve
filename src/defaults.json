{
  "init": {
    "headers": {
      "Accept": "application/json",
      "Content-Type": "application/json"
    },
    "method": "GET"
  },
  "opts": {
    "rawPayload": false,
    "responseAs": "json"
  }
}
