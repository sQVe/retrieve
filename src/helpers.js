import { compose } from './utility'

/**
 * createQuery :: {k:v} -> string
 */
export function createQuery(params) {
  if (!params) return ''

  const arr = Object.entries (params).map (([k, v]) =>
    [k, encodeURIComponent (v)].join ('=')
  )

  return arr.length ? arr.join ('&') : ''
}

/**
 * sanitizeUrl :: string -> string
 */
export function sanitizeUrl(url) {
  const addTrailingSlash = x => (x.endsWith ('/') ? x : `${x}/`)
  const removeLeadingSlash = x => (x.startsWith ('/') ? x.substring (1) : x)

  return url
    ? compose (
        addTrailingSlash,
        removeLeadingSlash
      ) (url)
    : ''
}
