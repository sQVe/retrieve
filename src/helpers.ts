import {
  Container,
  NotNil,
  OnResponse,
  QueryParams,
  ResolveAs,
  ResolveMethod,
} from './types'
import { isNonEmptyString } from './utility'

/** combineUrls :: [String] -> String */
export const combineUrls = (urls: (string | undefined)[]) => {
  const addTrailingSlash = (x: string) => (x.endsWith('/') ? x : `${x}/`)
  const removeLeadingSlash = (x: string) =>
    x.startsWith('/') ? x.substring(1) : x

  const validUrls = (urls || []).filter(isNonEmptyString)

  if (validUrls.length === 0) return ''
  if (validUrls.length === 1) return validUrls[0]

  return validUrls.reduce(
    (acc, url) => addTrailingSlash(acc) + removeLeadingSlash(url)
  )
}

/** createQuery :: {k: v} -> String */
export const createQuery = (params: QueryParams) => {
  const arr = Object.entries(params || {}).map(([k, v]) =>
    [k, encodeURIComponent(v)].join('=')
  )

  return arr.length ? arr.join('&') : ''
}

/** getResolveAsMethodName :: String -> String */
export const getResolveAsMethodName = (resolveAs?: ResolveAs) => {
  const methodNameLookup: { [index: string]: ResolveMethod } = {
    arraybuffer: 'arrayBuffer',
    blob: 'blob',
    formdata: 'formData',
    json: 'json',
    response: 'response',
    text: 'text',
  }

  if (resolveAs == null) return methodNameLookup.response

  const caseSafeResolveAs = resolveAs.toLowerCase()
  return methodNameLookup[caseSafeResolveAs] || resolveAs
}

/** handleResponse :: Response r => (r a -> void) -> r a -> void */
export const handleResponse = (onResponse?: OnResponse) => (res: Response) =>
  onResponse && onResponse(res)

/** preparePayload :: Container -> a -> b */
export const preparePayload = (container: Container) => (payload: NotNil) => {
  const { payloadAs } = container.options

  return payload != null && payloadAs != null ? payloadAs(payload) : payload
}

/** resolveResponse :: Response a -> String -> Promise b */
export const resolveResponse = (res: Response) => async (
  resolveAs?: ResolveAs
) => {
  const resolveAsMethodName = getResolveAsMethodName(resolveAs)

  if (resolveAsMethodName === 'response') return res
  if (resolveAsMethodName === 'json') {
    const text = await res.clone().text()

    if (text.length === 0) return undefined // eslint-disable-line fp/no-nil
    return res.json()
  }

  return res[resolveAsMethodName]()
}

/** combineContainers :: Container -> Container -> Container */
export const combineContainers = (a: Partial<Container>) => (
  b: Partial<Container>
): Container => ({
  url: combineUrls([a.url, b.url]),
  init: { ...(a.init || {}), ...(b.init || {}) },
  options: { ...(a.options || {}), ...(b.options || {}) },
})
