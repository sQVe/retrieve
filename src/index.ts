export { righttp, righttp as default } from './righttp'
