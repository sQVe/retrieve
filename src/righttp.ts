import { Container, Init, Options } from './types'
import { combineContainers, handleResponse, resolveResponse } from './helpers'
import { defaultContainer } from './constants'

/** request :: Container -> Promise a */
const request = async (container: Container) => {
  const {
    url,
    init,
    options: { onResponse, resolveAs },
  } = container

  if (url.length === 0)
    throw new Error('Righttp needs an URL to make a request.')

  const res = await fetch(url, init)

  handleResponse(onResponse)(res)

  if (!res.ok) throw new Error(`${res.status}: ${res.statusText}`)
  return resolveResponse(res)(resolveAs)
}

/**
 * righttp :: (String, Init, Options) -> {k: (Container -> Promise a)}
 *
 * TODO: Write proper JSDocs for `righttp` function and the functions it
 * returns.
 */
export function righttp(url: string, init: Init, options: Options) {
  const presetContainer = combineContainers(defaultContainer)({
    url,
    init,
    options,
  })
  const presetCombine = combineContainers(presetContainer)

  return {
    request: (url: string, init: Init, options: Options) =>
      request(presetCombine({ url, init, options })),
  }
}
