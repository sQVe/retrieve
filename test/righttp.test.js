const mockLastCombineContainerUnary = jest.fn()
const mockLastHandleResponseUnary = jest.fn()
const mockLastResolveResponseUnary = jest.fn()
const mockLastLoadPayloadUnary = jest.fn()

jest.mock('../src/helpers', () => {
  const helpers = jest.requireActual('../src/helpers')

  return {
    ...helpers,
    combineContainers: jest.fn(a =>
      mockLastCombineContainerUnary.mockImplementation(b =>
        helpers.combineContainers(a)(b)
      )
    ),
    handleResponse: jest.fn(a =>
      mockLastHandleResponseUnary.mockImplementation(b =>
        helpers.handleResponse(a)(b)
      )
    ),
    loadPayload: jest.fn(a =>
      mockLastLoadPayloadUnary.mockImplementation(b =>
        helpers.loadPayload(a)(b)
      )
    ),
    resolveResponse: jest.fn(res =>
      mockLastResolveResponseUnary.mockImplementation(resolveAs =>
        helpers.resolveResponse(
          resolveAs === 'FormData'
            ? // Mock formData method due to lack of support in node-fetch.
              { formData: () => resolveAs }
            : res
        )(resolveAs)
      )
    ),
  }
})

import mockJest from 'jest-fetch-mock'
import { Response } from 'cross-fetch'

import {
  combineContainers,
  handleResponse,
  loadPayload,
  resolveResponse,
} from '../src/helpers'
import { commonHttpStatuses, resolveAsMethodNameMap } from './setup/constants'
import { defaultContainer } from '../src/constants'
import { resolveAsResponses } from './setup/mocks'
import { righttp } from '../src/righttp'

mockJest.enableMocks()

describe('righttp', () => {
  beforeEach(fetch.resetMocks)
  afterEach(jest.clearAllMocks)

  describe('righttp', () => {
    const container = {
      url: 'foo',
      init: { bar: true },
      options: { baz: true },
    }

    it('should return an API', () => {
      const subject = righttp(...Object.values(container))
      const apiMethods = [
        'del',
        'get',
        'patch',
        'post',
        'preset',
        'put',
        'request',
      ]

      expect(Object.keys(subject).length).toEqual(apiMethods.length)
      apiMethods.forEach(key => {
        expect(subject[key]).toBeDefined()
      })
    })

    it('should set preset defaults for our container', () => {
      righttp()

      expect(combineContainers).toHaveBeenCalledTimes(2)
      expect(combineContainers).toHaveBeenCalledWith(defaultContainer)
    })

    it('should combine url, init and options', () => {
      const subject = righttp(...Object.values(container))

      expect(subject.preset).toMatchSnapshot()
      expect(combineContainers).toHaveBeenCalledTimes(2)
      expect(mockLastCombineContainerUnary).toHaveBeenCalledTimes(1)
      expect(mockLastCombineContainerUnary).toHaveBeenCalledWith(container)
    })

    describe('del', () => {
      it('should use our preset defaults', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp().del('foo')

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(fetch.mock.calls).toMatchSnapshot()
      })

      it('should preset a URL', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('preset').del('foo')
        const [firstCall] = fetch.mock.calls
        const [url] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(url).toBe('preset/foo')
      })

      it('should set DELETE as method in init', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('foo', { method: 'GET' }).del('bar')
        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.method).toBe('DELETE')
      })

      it('should load payload in init', async () => {
        expect.assertions(5)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('foo', {}, { payloadAs: x => x }).del(
          'bar',
          'baz'
        )
        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(loadPayload).toHaveBeenCalledTimes(1)
        expect(mockLastLoadPayloadUnary).toHaveBeenCalledWith('baz')
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.body).toBe('baz')
      })

      it('should not load payload in init when payload is nil', async () => {
        expect.assertions(7)
        fetch.mockResponse(resolveAsResponses.JSON)

        const firstSubject = await righttp(
          'foo',
          {},
          { payloadAs: x => x }
        ).del('bar')
        const secondSubject = await righttp(
          'foo',
          {},
          { payloadAs: x => x }
        ).del('bar', null)
        const [firstCall, secondCall] = fetch.mock.calls
        const [_, firstInit] = firstCall
        const [__, secondInit] = secondCall

        expect(firstSubject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(secondSubject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(loadPayload).toHaveBeenCalledTimes(2)
        expect(mockLastLoadPayloadUnary.mock.calls).toEqual([
          [undefined],
          [null],
        ])
        expect(fetch).toHaveBeenCalledTimes(2)
        expect(firstInit.body).toBeUndefined()
        expect(secondInit.body).toBeUndefined()
      })
    })

    describe('get', () => {
      it('should use our preset defaults', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp().get('foo')

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(fetch.mock.calls).toMatchSnapshot()
      })

      it('should preset a URL', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('preset').get('foo')
        const [firstCall] = fetch.mock.calls
        const [url] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(url).toBe('preset/foo')
      })

      it('should set GET as method in init', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('foo', { method: 'POST' }).get('bar')
        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.method).toBe('GET')
      })

      it('should append query params to URL', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp().get('foo', {
          bar: 'baz',
          qux: 123,
        })
        const [firstCall] = fetch.mock.calls
        const [url] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(url).toBe('foo?bar=baz&qux=123')
      })
    })

    describe('patch', () => {
      it('should use our preset defaults', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp().patch('foo')

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(fetch.mock.calls).toMatchSnapshot()
      })

      it('should preset a URL', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('preset').patch('foo')
        const [firstCall] = fetch.mock.calls
        const [url] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(url).toBe('preset/foo')
      })

      it('should set PATCH as method in init', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('foo', { method: 'GET' }).patch('bar')
        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.method).toBe('PATCH')
      })

      it('should load payload in init', async () => {
        expect.assertions(5)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('foo', {}, { payloadAs: x => x }).patch(
          'bar',
          'baz'
        )
        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(loadPayload).toHaveBeenCalledTimes(1)
        expect(mockLastLoadPayloadUnary).toHaveBeenCalledWith('baz')
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.body).toBe('baz')
      })
    })

    describe('post', () => {
      it('should use our preset defaults', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp().post('foo')

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(fetch.mock.calls).toMatchSnapshot()
      })

      it('should preset a URL', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('preset').post('foo')
        const [firstCall] = fetch.mock.calls
        const [url] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(url).toBe('preset/foo')
      })

      it('should set PATCH as method in init', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('foo', { method: 'GET' }).post('bar')
        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.method).toBe('POST')
      })

      it('should load payload in init', async () => {
        expect.assertions(5)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('foo', {}, { payloadAs: x => x }).post(
          'bar',
          'baz'
        )
        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(loadPayload).toHaveBeenCalledTimes(1)
        expect(mockLastLoadPayloadUnary).toHaveBeenCalledWith('baz')
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.body).toBe('baz')
      })
    })

    describe('put', () => {
      it('should use our preset defaults', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp().put('foo')

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(fetch.mock.calls).toMatchSnapshot()
      })

      it('should preset a URL', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('preset').put('foo')
        const [firstCall] = fetch.mock.calls
        const [url] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(url).toBe('preset/foo')
      })

      it('should set PATCH as method in init', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('foo', { method: 'GET' }).put('bar')
        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.method).toBe('PUT')
      })

      it('should load payload in init', async () => {
        expect.assertions(5)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('foo', {}, { payloadAs: x => x }).put(
          'bar',
          'baz'
        )
        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(loadPayload).toHaveBeenCalledTimes(1)
        expect(mockLastLoadPayloadUnary).toHaveBeenCalledWith('baz')
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.body).toBe('baz')
      })
    })

    describe('request', () => {
      it('should use our preset defaults', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp().request('foo')

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(fetch.mock.calls).toMatchSnapshot()
      })

      it('should preset a URL', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('preset').request('foo')
        const [firstCall] = fetch.mock.calls
        const [url] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(url).toBe('preset/foo')
      })

      it('should preset init', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp(undefined, { method: 'POST' }).request(
          'foo'
        )
        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.method).toBe('POST')
      })

      it('should preset options', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.Text)

        const subject = await righttp(undefined, undefined, {
          resolveAs: 'Text',
        }).request('foo')

        expect(subject).toBe(resolveAsResponses.Text)
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(mockLastResolveResponseUnary).toHaveBeenCalledWith('Text')
      })

      it('should make a request to specified URL', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp('foo/bar').request('baz/qux')
        const [firstCall] = fetch.mock.calls
        const [url] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(url).toBe('foo/bar/baz/qux')
      })

      it('should make a request with custom method in init', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp().request('foo', { method: 'POST' })
        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.method).toBe('POST')
      })

      it('should make a request with custom headers in init', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.JSON)

        const subject = await righttp().request('foo', {
          headers: { bar: 'baz' },
        })

        const [firstCall] = fetch.mock.calls
        const [_, init] = firstCall

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(init.headers).toEqual({ bar: 'baz' })
      })

      it('should make a request with custom options', async () => {
        expect.assertions(3)
        fetch.mockResponse(resolveAsResponses.Text)

        const subject = await righttp().request('foo', undefined, {
          resolveAs: 'Text',
        })

        expect(subject).toBe(resolveAsResponses.Text)
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(mockLastResolveResponseUnary).toHaveBeenCalledWith('Text')
      })

      it('should call handleResponse with onResponse and response', async () => {
        expect.assertions(4)
        fetch.mockResponse(resolveAsResponses.JSON)

        const onResponse = () => {}
        const subject = await righttp().request('foo', undefined, {
          onResponse,
        })

        expect(subject).toEqual(JSON.parse(resolveAsResponses.JSON))
        expect(handleResponse).toHaveBeenCalledTimes(1)
        expect(handleResponse).toHaveBeenCalledWith(onResponse)
        expect(mockLastHandleResponseUnary.mock.calls).toMatchSnapshot()
      })

      Object.entries(resolveAsResponses).forEach(([resolveAs, res]) => {
        it(`should resolve request as ${resolveAs} when resolveAs option is ${resolveAs}`, async () => {
          expect.assertions(5)

          const methodName = resolveAsMethodNameMap[resolveAs]

          fetch.mockResponse(res)

          const subject = await righttp(undefined, undefined, {
            resolveAs,
          }).request('foo')

          expect(subject).toEqual(await new Response(res)[methodName]())
          expect(fetch).toHaveBeenCalledTimes(1)
          expect(resolveResponse).toHaveBeenCalledTimes(1)
          expect(resolveResponse.mock.calls).toMatchSnapshot()
          expect(mockLastResolveResponseUnary).toHaveBeenCalledWith(resolveAs)
        })
      })

      it('should resolve request as FormData when resolveAs option is FormData', async () => {
        expect.assertions(5)

        const formData = new FormData()
        formData.append('form', 'Data')

        fetch.mockResponse(formData)

        const subject = await righttp(undefined, undefined, {
          resolveAs: 'FormData',
        }).request('foo')

        expect(subject).toBe('FormData')
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(resolveResponse).toHaveBeenCalledTimes(1)
        expect(resolveResponse.mock.calls).toMatchSnapshot()
        expect(mockLastResolveResponseUnary).toHaveBeenCalledWith('FormData')
      })

      it('should resolve request as Response when resolveAs option is Response', async () => {
        expect.assertions(6)

        fetch.mockResponse(resolveAsResponses.Text)

        const subject = await righttp(undefined, undefined, {
          resolveAs: 'Response',
        }).request('foo')

        expect(subject).toMatchSnapshot()
        expect(subject.text()).resolves.toBe(resolveAsResponses.Text)
        expect(fetch).toHaveBeenCalledTimes(1)
        expect(resolveResponse).toHaveBeenCalledTimes(1)
        expect(resolveResponse.mock.calls).toMatchSnapshot()
        expect(mockLastResolveResponseUnary).toHaveBeenCalledWith('Response')
      })

      it('should throw when given empty URL', () => {
        expect.assertions(1)

        expect(righttp().request()).rejects.toThrowError(
          new Error('Righttp needs an URL to make a request.')
        )
      })

      it('should throw when failing to fetch', () => {
        expect.assertions(1)

        const fetchRejectError = new TypeError('Fetch rejected.')

        fetch.mockReject(fetchRejectError)

        expect(righttp().request('foo')).rejects.toThrowError(fetchRejectError)
      })

      it('should throw when fetch is aborted', () => {
        expect.assertions(1)

        fetch.mockAbort()

        expect(righttp().request('foo')).rejects.toThrow()
      })

      it('should throw when response is unsuccessful', () => {
        const commonUnsuccessfulHttpStatuses = commonHttpStatuses.filter(
          ({ status }) => status > 299
        )

        expect.assertions(commonUnsuccessfulHttpStatuses.length)

        commonUnsuccessfulHttpStatuses.forEach(
          async ({ status, statusText }) => {
            fetch.mockResponse('', { status })

            expect(righttp().request('foo')).rejects.toThrowError(
              new Error(`${status}: ${statusText}`)
            )
          }
        )
      })
    })
  })
})
