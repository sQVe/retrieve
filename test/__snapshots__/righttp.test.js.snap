// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`righttp righttp request should call handleResponse with onResponse and response 1`] = `
Array [
  Array [
    Response {
      "size": 0,
      "timeout": 0,
      Symbol(Body internals): Object {
        "body": Object {
          "data": Array [
            91,
            34,
            106,
            115,
            111,
            110,
            34,
            93,
          ],
          "type": "Buffer",
        },
        "disturbed": true,
        "error": null,
      },
      Symbol(Response internals): Object {
        "counter": undefined,
        "headers": Headers {
          Symbol(map): Object {
            "Content-Type": Array [
              "text/plain;charset=UTF-8",
            ],
          },
        },
        "status": 200,
        "statusText": "OK",
        "url": undefined,
      },
    },
  ],
]
`;

exports[`righttp righttp request should make a request with our defaults 1`] = `
Array [
  Array [
    "foo/bar",
    Object {
      "headers": Object {
        "Accept": "application/json",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`righttp righttp request should resolve request as ArrayBuffer when resolveAs option is ArrayBuffer 1`] = `
Array [
  Array [
    Response {
      "size": 0,
      "timeout": 0,
      Symbol(Body internals): Object {
        "body": Object {
          "data": Array [
            97,
            114,
            114,
            97,
            121,
            66,
            117,
            102,
            102,
            101,
            114,
          ],
          "type": "Buffer",
        },
        "disturbed": true,
        "error": null,
      },
      Symbol(Response internals): Object {
        "counter": undefined,
        "headers": Headers {
          Symbol(map): Object {},
        },
        "status": 200,
        "statusText": "OK",
        "url": undefined,
      },
    },
  ],
]
`;

exports[`righttp righttp request should resolve request as Blob when resolveAs option is Blob 1`] = `
Array [
  Array [
    Response {
      "size": 0,
      "timeout": 0,
      Symbol(Body internals): Object {
        "body": Object {
          "data": Array [
            91,
            111,
            98,
            106,
            101,
            99,
            116,
            32,
            66,
            108,
            111,
            98,
            93,
          ],
          "type": "Buffer",
        },
        "disturbed": true,
        "error": null,
      },
      Symbol(Response internals): Object {
        "counter": undefined,
        "headers": Headers {
          Symbol(map): Object {
            "Content-Type": Array [
              "text/plain;charset=UTF-8",
            ],
          },
        },
        "status": 200,
        "statusText": "OK",
        "url": undefined,
      },
    },
  ],
]
`;

exports[`righttp righttp request should resolve request as FormData when resolveAs option is FormData 1`] = `
Array [
  Array [
    Response {
      "size": 0,
      "timeout": 0,
      Symbol(Body internals): Object {
        "body": Object {
          "data": Array [
            91,
            111,
            98,
            106,
            101,
            99,
            116,
            32,
            70,
            111,
            114,
            109,
            68,
            97,
            116,
            97,
            93,
          ],
          "type": "Buffer",
        },
        "disturbed": false,
        "error": null,
      },
      Symbol(Response internals): Object {
        "counter": undefined,
        "headers": Headers {
          Symbol(map): Object {
            "Content-Type": Array [
              "text/plain;charset=UTF-8",
            ],
          },
        },
        "status": 200,
        "statusText": "OK",
        "url": undefined,
      },
    },
  ],
]
`;

exports[`righttp righttp request should resolve request as JSON when resolveAs option is JSON 1`] = `
Array [
  Array [
    Response {
      "size": 0,
      "timeout": 0,
      Symbol(Body internals): Object {
        "body": Object {
          "data": Array [
            91,
            34,
            106,
            115,
            111,
            110,
            34,
            93,
          ],
          "type": "Buffer",
        },
        "disturbed": true,
        "error": null,
      },
      Symbol(Response internals): Object {
        "counter": undefined,
        "headers": Headers {
          Symbol(map): Object {
            "Content-Type": Array [
              "text/plain;charset=UTF-8",
            ],
          },
        },
        "status": 200,
        "statusText": "OK",
        "url": undefined,
      },
    },
  ],
]
`;

exports[`righttp righttp request should resolve request as Response when resolveAs option is Response 1`] = `
Response {
  "size": 0,
  "timeout": 0,
  Symbol(Body internals): Object {
    "body": Object {
      "data": Array [
        116,
        101,
        120,
        116,
      ],
      "type": "Buffer",
    },
    "disturbed": false,
    "error": null,
  },
  Symbol(Response internals): Object {
    "counter": undefined,
    "headers": Headers {
      Symbol(map): Object {
        "Content-Type": Array [
          "text/plain;charset=UTF-8",
        ],
      },
    },
    "status": 200,
    "statusText": "OK",
    "url": undefined,
  },
}
`;

exports[`righttp righttp request should resolve request as Response when resolveAs option is Response 2`] = `
Array [
  Array [
    Response {
      "size": 0,
      "timeout": 0,
      Symbol(Body internals): Object {
        "body": Object {
          "data": Array [
            116,
            101,
            120,
            116,
          ],
          "type": "Buffer",
        },
        "disturbed": true,
        "error": null,
      },
      Symbol(Response internals): Object {
        "counter": undefined,
        "headers": Headers {
          Symbol(map): Object {
            "Content-Type": Array [
              "text/plain;charset=UTF-8",
            ],
          },
        },
        "status": 200,
        "statusText": "OK",
        "url": undefined,
      },
    },
  ],
]
`;

exports[`righttp righttp request should resolve request as Text when resolveAs option is Text 1`] = `
Array [
  Array [
    Response {
      "size": 0,
      "timeout": 0,
      Symbol(Body internals): Object {
        "body": Object {
          "data": Array [
            116,
            101,
            120,
            116,
          ],
          "type": "Buffer",
        },
        "disturbed": true,
        "error": null,
      },
      Symbol(Response internals): Object {
        "counter": undefined,
        "headers": Headers {
          Symbol(map): Object {
            "Content-Type": Array [
              "text/plain;charset=UTF-8",
            ],
          },
        },
        "status": 200,
        "statusText": "OK",
        "url": undefined,
      },
    },
  ],
]
`;

exports[`righttp righttp should combine url, init and options 1`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "init": Object {
        "bar": true,
        "headers": Object {
          "Accept": "application/json",
          "Content-Type": "application/json",
        },
        "method": "GET",
      },
      "options": Object {
        "baz": true,
        "payloadAs": [Function],
        "resolveAs": "JSON",
      },
      "url": "foo",
    },
  },
]
`;
